src/cascadia/TerminalApp/TerminalPage.cpp

```diff
@@ -137,51 +146,79 @@ namespace winrt::TerminalApp::implementation
+        _layoutUpdatedRevoker = _tabContent.LayoutUpdated(winrt::auto_revoke, { this, &TerminalPage::_OnFirstLayout });
+    }
+
+    // Method Description:
+    // - This method is caled once on startup, on the first LayoutUpdated event.
+    //   We'll use this event to know that we have an ActualWidth and
+    //   ActualHeight, so we can now attempt to process our list of startup
+    //   actions.
+    // - We'll remove this event handler when the event is first handled.
+    // - If there are no startup actions, we'll open a single tab with the
+    //   default profile.
+    // Arguments:
+    // - <unused>
+    // Return Value:
+    // - <none>
+    void TerminalPage::_OnFirstLayout(const IInspectable& /*sender*/, const IInspectable& /*eventArgs*/)
+    {
+        // Only let this succeed once.
+        _layoutUpdatedRevoker.revoke();
+
+        // This event fires every time the layout changes, but it is always the
+        // last one to fire in any layout change chain. That gives us great
+        // flexibility in finding the right point at which to initialize our
+        // renderer (and our terminal). Any earlier than the last layout update
+        // and we may not know the terminal's starting size.
+        if (_startupState == StartupState::NotInitialized)
         {
+            _startupState = StartupState::InStartup;
             _appArgs.ValidateStartupCommands();
-
-            // This will kick off a chain of events to perform each startup
-            // action. As each startup action is completed, the next will be
-            // fired.
-            _ProcessNextStartupAction();
+            if (_appArgs.GetStartupActions().empty())
+            {
+                _OpenNewTab(nullptr);
+                _startupState = StartupState::Initialized;
+                _InitializedHandlers(*this, nullptr);
+            }
+            else
+            {
+                _ProcessStartupActions();
+            }
         }
     }
 
     // Method Description:
-    // - Process the next startup action in our list of startup actions. When
-    //   that action is complete, fire the next (if there are any more).
+    // - Process all the startup actions in our list of startup actions. We'll
+    //   do this all at once here.
     // Arguments:
     // - <none>
     // Return Value:
     // - <none>
-    fire_and_forget TerminalPage::_ProcessNextStartupAction()
+    fire_and_forget TerminalPage::_ProcessStartupActions()
     {
         // If there are no actions left, do nothing.
         if (_appArgs.GetStartupActions().empty())
         {
             return;
         }
-
-        // Get the next action to be processed
-        auto nextAction = _appArgs.GetStartupActions().front();
-        _appArgs.GetStartupActions().pop_front();
-
         auto weakThis{ get_weak() };
 
         // Handle it on the UI thread.
         co_await winrt::resume_foreground(Dispatcher(), CoreDispatcherPriority::Low);
```

Comment:
We may want to get onto the normal priority queue _or_ make this no longer a coroutine since LayoutUpdated comes in on the dispatcher queue already

Reply:
if we get if off the queue entirely, we can dispense with all of the weak stuff _and_ the `_appArgs` check (because iterating over an empty list will be a noop)

Reply:
Curiously, it doesn't work if you do it immediately on _this_ pass of the UI thread. I will bump it up to Normal though